verifier/src/evidence.rs

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+//! The full set of evidence needed for attesting a quote
+
+use crate::{Accessor, Advisories, Error, Result, TcbInfo, TcbInfoRaw};
+use der::DecodePem;
+use mc_sgx_core_types::{
+    Attributes, ConfigId, ConfigSvn, CpuSvn, ExtendedProductId, FamilyId, IsvProductId, IsvSvn,
+    MiscellaneousSelect, MrEnclave, MrSigner, ReportData,
+};
+use mc_sgx_dcap_types::{CertificationData, Quote3, TcbInfo as QuoteTcbInfo};
+use x509_cert::Certificate;
+
+/// The full set of evidence needed for verifying a quote
+///
+/// A wrapping container for a `Quote3` and a `TcbInfoRaw`. This can be used
+/// with the majority of the `Verifier` implementations from this crate.
+/// This allows one to compose one verifier and use an [`Evidence`] instance in
+/// the `verify()` method.
+///
+/// Importantly this will derive the [`Advisories`] related to the provided
+/// `quote` and `tcb_info_raw`, so that one can verify the allowed advisories.
+#[derive(Debug)]
+pub struct Evidence<'a, Q> {
+    quote: Quote3<Q>,
+    tcb_info_raw: TcbInfoRaw<'a>,
+    advisories: Advisories,
+}
+
+impl<'a, Q: AsRef<[u8]>> Evidence<'a, Q> {
+    /// Create a new instance
+    pub fn new(quote: Quote3<Q>, tcb_info_raw: TcbInfoRaw<'a>) -> Result<Self> {
+        let quote_tcb_info = tcb_info_try_from_quote(&quote)?;
+        let tcb_info = TcbInfo::try_from(&tcb_info_raw)?;
+        let advisories = tcb_info.advisories(&quote_tcb_info)?;
+        Ok(Self {
+            quote,
+            tcb_info_raw,
+            advisories,
+        })
+    }
+}
+
+impl<'a, Q> Accessor<TcbInfoRaw<'a>> for Evidence<'a, Q> {
+    fn get(&self) -> TcbInfoRaw<'a> {
+        self.tcb_info_raw.clone()
+    }
+}
+
+impl<'a, Q: Clone> Accessor<Quote3<Q>> for Evidence<'a, Q> {
+    fn get(&self) -> Quote3<Q> {
+        self.quote.clone()
+    }
+}
+
+impl<'a, Q> Accessor<Advisories> for Evidence<'a, Q> {
+    fn get(&self) -> Advisories {
+        self.advisories.clone()
+    }
+}
+
+// TODO think this should go in tcb.rs of `mc-sgx-dcap-types`
+fn tcb_info_try_from_quote<Q: AsRef<[u8]>>(quote: &Quote3<Q>) -> Result<QuoteTcbInfo> {
+    let signature_data = quote.signature_data();
+    let certification_data = signature_data.certification_data();
+    let CertificationData::PckCertificateChain(pem_chain) = certification_data else {
+        return Err(Error::UnsupportedQuoteCertificationData);
+    };
+    let leaf_cert = pem_chain
+        .into_iter()
+        .next()
+        .ok_or(Error::UnsupportedQuoteCertificationData)?;
+    let certificate = Certificate::from_pem(leaf_cert)?;
```

Comment:
This feels a bit clunky to me.
Right now the `Quote3` contains certification data and that has different variants, but I'm wondering if there should be different quote variants.
This is more or less a remote attestation quote.
There are also quotes for local attestation, between enclaves on the same machine. These don't need to call out to the internet and use type 3, from https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf
> 3: Byte array that contains
concatenation of PPID encrypted using
RSA-3072-OAEP, CPUSVN, PCESVN (LE),
PCEID (LE).

Reply:
I think having separate types would allow us to leverage the type system for behavior customization

The following is an oversimplification from https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf
``` 
 ┌───────────────────────────────────┐
 │                                   │
 │          Quote Header             │
 │                                   │
 ├───────────────────────────────────┤
 │                                   │
 │       Quoting Enclave Report      │
 │                                   │
 ├───────────────────────────────────┤
 │                                   │
 │       Certification Data          │
 │(Varies based on quote generation) │
 │                                   │
 └───────────────────────────────────┘
```
Right now the client gets a quote and then needs to get and try to process the certification data.

If we had a RemoteAttestationQuote, we could ensure that the Certification data was correct and thus remove the need for the clients to handle possible errors, they would instead fail to get a quote. Also the clients would be able to call `leaf_cert()` and or `cert_chain()` without digging down to the certification data and matching on it.

In the end all this may just be polish that isn't that beneficial

Reply:
Okay, that makes sense. Maybe a RemoteAttestationQuote that has a TryFrom<Quote3> and extracts the certification data at that point would be nice, but it doesn't seem like a huge deal either way.